File: src/misc/math/game/GameRect.java
package misc.math.game;

import misc.Debug;
import misc.math.game.GameVector;
import misc.math.game.GamePosition;
import misc.math.game.GameSize;

public class GameRect
{
	private GamePosition position;
	private GameSize size;

	public GameRect(GamePosition position, GameSize size)
	{
		this.position = new GamePosition(position);
		this.size = new GameSize(size);
	}

	public GameRect(GameRect rect)
	{
		this(rect.getPosition(), rect.getSize());
	}

	public GameRect(float x, float y, float w, float h)
	{
		this(new GamePosition(x, y), new GameSize(w, h));
	}

	public GameRect(GameVector p1, GameVector p2)
	{
		position = new GamePosition(p1);
		size = new GameSize(p1.minus(p2));
	}

	public GameRect()
	{
		this(new GamePosition(0,0), new GameSize(0,0));
	}

	public boolean equals(GameRect rect)
	{
		return (getPosition().equals(rect.getPosition()) && getSize().equals(rect.getSize()));
	}

	public void setPosition(GameVector point) { position = new GamePosition(point); }
	public void setPosition(float x, float y) { position = new GamePosition(x, y); }
	public void setSize(GameVector point) { size = new GameSize(point); }
	public void setSize(float x, float y) { size = new GameSize(x, y); }

	public final float getBot() { return (getY() + getHeight()/2); }
	public final float getLeft() { return getX() - getWidth()/2; }
	public final float getRight() { return (getX() + getWidth()/2); }
	public final float getTop() { return getY() - getHeight()/2; }

	public GamePosition getPosition() { return position; }
	public float getX() { return getPosition().getX(); }
	public float getY() { return getPosition().getY(); }

	public GameSize getSize() { return size; }
	public float getWidth() { return getSize().getX(); }
	public float getHeight() { return getSize().getY(); }
}

File: cpp/src/math/game/GameRect.cpp
#include "GameRect.hpp"

GameRect::GameRect(float x, float y, float w, float h) : position(x, y), size(w, h)
{}

GameRect::GameRect(const GameRect& rect) : position(rect.getPosition()), size(rect.getSize())
{}

GameRect::GameRect(const GameVector& position, const GameVector& size) : position(position), size(size)
{}

GameVector GameRect::getPosition() const
{
	return position;
}

GameVector GameRect::getSize() const
{
	return size;
}

float GameRect::getLeft() const
{
	return getPosition().x;
}

float GameRect::getRight() const
{
	return getPosition().x + getSize().x;
}

float GameRect::getTop() const
{
	return getPosition().y;
}

float GameRect::getBot() const
{
	return getPosition().y + getSize().y;
}

float GameRect::getWidth() const { return getRight() - getLeft(); }
float GameRect::getHeight() const { return getBot() - getTop(); }

bool GameRect::collidesWith(const GameRect& r) const
{
	return     getLeft() <= r.getRight()
		&& getRight() >= r.getLeft()
		&& getTop() <= r.getBot()
		&& getBot() >= r.getTop();
}

bool GameRect::isValid() const
{
	return ((getLeft() < getRight()) && (getTop() < getBot()));
}

bool GameRect::operator==(const GameRect& rect)
{
	return rect.getPosition() == getPosition() && rect.getSize() == getSize();
}

std::string GameRect::toString() const
{
	return "GR(" + getPosition().toString() + ", " + getSize().toString() + ")";
}

GameRect GameRect::getIntersectionRect(const GameRect& r, const GameRect& s)
{
	return GameRect(std::max(r.getLeft(), s.getLeft()), std::min(r.getRight(), s.getRight()), std::max(r.getTop(), s.getTop()), std::min(r.getBot(), s.getBot()));
}
